### ISSUE TITLE
offset(-1, offset(1, ...)) produces "island" artifacts

### ISSUE THREAD

**Issue by paulftw** [jscad/OpenJSCAD.org#1017]
## Steps to reproduce:

```
const offset = (delta, c) =>
  expansions.offset({ delta, corners: 'round', segments: 16 }, c)

export function main() {
  return [
    offset(-1, offset(1, primitives.square(4))),

    transforms.translate(
      [0, 3, 0],
      offset(-0.9, offset(1, primitives.square(4))),
    ),
  ]
}
```

[This produces](https://scriptcad.com/paulftw/jscad-tests) the following:
<img width="563" alt="Screen Shot 2022-02-28 at 2 10 35 PM" src="https://user-images.githubusercontent.com/769142/155981511-ebacc0c3-2208-4805-bf36-48dff19999b6.png">

Segment of a circle should **not** be there.

**Comment by paulftw** [comment]
Artifacts are visually similar to #907 but that one doesn't seem to use `offset`

**Comment by z3dev** [comment]
@paulftw cool examples.

from looking at this, the last round segments may be illformed after the first offset.

**Comment by platypii** [comment]
Notes:
1. The way that rounded edges are constructed results in the two "side" edges coming together and eating the "middle" edge. This is because the outer edge angle is less than the middle edge angles. Not a bug, just how the construction works:

Offset delta = -0.92:
![offset-0 92](https://user-images.githubusercontent.com/1766297/156089109-f2ffe090-74ff-4fc1-82df-3a092ea9f212.png)

Offset delta = -1:
![offset-1](https://user-images.githubusercontent.com/1766297/156089114-1d0a7467-5174-4dd3-a986-642e225afddc.png)

2. Regardless of how the shape was constructed, it's a valid geometry that someone might want to offset. The problem seems to be that the code in `offsetFromPoints` recognizes that the middle edge needs to be removed, but fails to find the intersection point of the remaining side edges.
3. The path closure is handled differently, and that is the reason that the artifact only appears on the bottom left corner of the square, not every corner.

Continuing to investigate...

**Comment by z3dev** [comment]
@platypii take a look at line 66… I’m wondering if there’s a special condition that requires additional math to find the correct point of intersection.

**Comment by z3dev** [comment]
Yeah. Those little ‘middle’ edges will be discarded on normal corners because that edge occurs in the middle of outer segments. The normal loops removes those ‘middle’ edges. For closure to work properly, another iteration / check should be performed to adjust the outer segments if intersecting.

**Comment by z3dev** [comment]
I’ll read the paper.

I think #782 has the same issue, where the shape contracts upon itself, and creates edges with inverted angles.

**Comment by z3dev** [comment]
I like this note…

> However, two kind of events might ocurr which do change the topology of the offset polygon; that is, after such events, the offset polygons hsa a different set of connected edges..

Maybe one suggestion is to throw an error if the topology changes, i.e. something bad happened.

**Comment by z3dev** [comment]
By the way, returning two or more pieces (outlines) should be a error. The offset of a line is exactly that… a line… it can cross… it can be strange… it can be wrong. But it’s still a line.

Geometry 2D is another story.

**Comment by z3dev** [comment]
This is far more complex then initially coded. 😃

**Comment by paulftw** [comment]
Random idea: negative offset of a N vertices polygon is boolean difference of the polygon and N rectangles, where rectangle `i` is parallel to the edge `i`, has its length and width equal to `delta` (bad explanation but I think you get the idea).
positive offset is boolean union of the input polygon and similar rectangles.
Assuming boolean operations work better than offset and aren't too expensive (they use BSP, don't they?), we can reduce this difficult problem to an already solved one.
